server.js

```diff
@@ -3,33 +3,99 @@ import bodyParser from 'body-parser'
-// import avocadoSalesData from './data/avocado-sales.json'
-// import booksData from './data/books.json'
-// import netflixData from './data/netflix-titles.json'
-// import topMusicData from './data/top-music.json'
+import booksData from './data/books.json'
+import Book from './models/book'
 
+//connecting to the database
 const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/project-mongo"
 mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
 mongoose.Promise = Promise
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
-const port = process.env.PORT || 8080
+const port = process.env.PORT || 8081
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
 app.use(cors())
 app.use(bodyParser.json())
 
-// Start defining your routes here
+//if the database working- catch if not
+//if next isn't invoked then express won't call the next line of code.
+//use mongoose ready state. if 1 then connected
+app.use((req, res, next) => {
+  if(mongoose.connection.readyState === 1){
+    next()
+  } else {
+    res.status(503).json({ error: 'service unavailable' })
+  }
+})
+
+//cleans database before population
+if (process.env.RESET_DATABASE) {
+  const populateDatabase = async () => {
+    await Book.deleteMany()
+
+    booksData.forEach(book => {
+      const newBook = new Book(book)
+      newBook.save()
+    })
+  }
+  populateDatabase()
+}
+
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send('Hello again world')
+})
+
+//limit is amount of search results for a single page
+//i stands for ignore case -->in regexp
+//math.ceil rounds the number up to the next largest integer.
+app.get('/books', async (req, res) => {
+  const { title, author, rating, page = 1, limit = 10 } = req.query
+  const titleRegex = new RegExp(title, 'i')
+  const authorRegex = new RegExp(author, 'i')
+
+  try {
+    const sortBooks = (rating) => {
+      if (rating === 'high') {
+        return {average_rating: - 1}
+      } else if (rating === 'low') {
+        return {average_rating: 1}
+      }
+    }
+
+    const books = await Book.find({
+      title: titleRegex,
+      authors: authorRegex
+    })
+      .sort(sortBooks(rating))
+      .limit(limit * 1)
+      .skip((page - 1) * limit)
+      .exec()
+    
+    //gets total amount of documents in the Book collection
+    //totalPages - takes the above and divides it with the limit(nr of results per page) to get the total amount of pages
+    const count = await Book.countDocuments()
+    res.json({
+      books,
+      totalPages: Math.ceil(count / limit),
+      currentPage: page
+    })
+  } catch(err) {
+    res.status(404).json({ error: 'books not found'})
+  }
+})
+
+//findOne returns one element. returns an object instead of an array with an object
+app.get('/books/book/:id', async (req, res) => {
```

Comment:
`id` is a default key for this document so better endpoint URL would be `/books/:id`